File: problems/045-slowest-k-cases/solutions/python/main.py
import heapq
import sys


tokens = sys.stdin.buffer.read().split()
n = int(tokens[0])
k = int(tokens[1])
heap: list[tuple[int, int]] = []
output = []

for index in range(1, n + 1):
    cost = int(tokens[index + 1])
    candidate = (cost, -index)
    if len(heap) < k:
        heapq.heappush(heap, candidate)
    elif candidate > heap[0]:
        heapq.heapreplace(heap, candidate)
    if index >= k:
        output.append(f"{-heap[0][1]} {heap[0][0]}\n")

sys.stdout.write("".join(output))

File: problems/045-slowest-k-cases/solutions/rust/main.rs
use std::cmp::Ordering;
use std::collections::BinaryHeap;
use std::io::{self, BufWriter, Read, Write};

#[derive(Clone, Copy, Eq, PartialEq)]
struct Case {
    cost: u64,
    index: usize,
}

impl Ord for Case {
    fn cmp(&self, other: &Self) -> Ordering {
        other
            .cost
            .cmp(&self.cost)
            .then_with(|| self.index.cmp(&other.index))
    }
}

impl PartialOrd for Case {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn better(left: Case, right: Case) -> bool {
    left.cost > right.cost || (left.cost == right.cost && left.index < right.index)
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut tokens = input.split_whitespace();
    let n: usize = tokens.next().unwrap().parse().unwrap();
    let k: usize = tokens.next().unwrap().parse().unwrap();
    let mut heap = BinaryHeap::with_capacity(k);
    let mut output = BufWriter::new(io::stdout().lock());
    for index in 1..=n {
        let cost: u64 = tokens.next().unwrap().parse().unwrap();
        let candidate = Case { cost, index };
        if heap.len() < k {
            heap.push(candidate);
        } else if better(candidate, *heap.peek().unwrap()) {
            heap.pop();
            heap.push(candidate);
        }
        if index >= k {
            let answer = heap.peek().unwrap();
            writeln!(output, "{} {}", answer.index, answer.cost).unwrap();
        }
    }
}
